core/src/main/java/org/apache/accumulo/fate/zookeeper/ZooCache.java

```diff
@@ -402,19 +426,37 @@ public T retry() {
          * a special case that looks for Code.NONODE in the KeeperException, then non-existence can
          * not be cached.
          */
+
+        Matcher configMatcher = TABLE_SETTING_CONFIG_PATTERN.matcher(zPath);
+        boolean createWatch = !configMatcher.matches();
+
         cacheWriteLock.lock();
         try {
           final ZooKeeper zooKeeper = getZooKeeper();
-          Stat stat = zooKeeper.exists(zPath, watcher);
+          Stat stat = zooKeeper.exists(zPath, createWatch ? watcher : null);
           byte[] data = null;
           if (stat == null) {
             if (log.isTraceEnabled()) {
               log.trace("zookeeper did not contain {}", zPath);
             }
           } else {
             try {
-              data = zooKeeper.getData(zPath, watcher, stat);
+              data = zooKeeper.getData(zPath, createWatch ? watcher : null, stat);
               zstat = new ZcStat(stat);
+              // Create the first watcher for /accumulo/{InstanceID/table-config-version
+              // so that the next modification of a table config in ZooUtil.putData
+              // will trigger a watch event for that  znode.
+              // Watchers need to be reset on a node if a watch event was triggered.
+              // The reset of the watcher for the above mentioned node will be done in the NodeDataChanged
+              // condition of the ZCacheWatcher.process function when it has been triggered
+              // by a change created in ZooUtil.putData.
+              if (!tableConfigWatcherSet) {
```

Comment:
I never see this set back to false once the watcher was triggered.

Reply:
You don't want it to set it back to false.  This just gets the Watch on the "table-config-version" node the first time.  I am trying to find a better way to put the watch on this node the first time.

Reply:
But I thought you needed to reset the watcher every time it was triggered?

Reply:
Yes,  the watcher for the table-config-version node gets reset in the process method when the NodeDataChanged event gets called by the watch on the table-config-version node due to its modification inside of ZooUtil.putData.  

No table config znodes are ever watched. That is the specification.